Synchronize definitions and timeline saving

arnaudvergnet opened the pull request:
Definitions and current timeline are kept in memory inside a singleton, and saved to disk when needed.

This data is only saved to the config file on timeline end/start.

While in the editor, default values are loaded as I can't seem to make the saves singleton load.

User can request a definitions reset either by a dedicated function, or by passing an argument to the start method.

Current timeline is saved so user can read it while running or load it back on next run. This feature combined with definitions saving makes a complete save/restore system, in which users can exit mid-timeline without negative effects.

I also added a lot of utility functions in `DialogicClass.gd`, to allow the user to edit and query variables and glossary definitions. It only repeats functions in the singleton, but this way the user only has one class to worry about and it prevents him from accessing unsafe methods.

Closes #102

coppolaemilio commented on addons/dialogic/Other/DialogicDefinitionsUtil.gd:
```diff
@@ -0,0 +1,55 @@
+	config.set_value(section, 'value', str(value))
+
+
+static func set_definition_glossary(config: ConfigFile, section: String, name: String,  extra_title: String,  extra_text: String,  extra_extra: String):
```
While reviewing the code I noticed that we could maybe save this glossary type variables as an array on the `value` key instead of creating the three `extra_title`, `extra_text` and `extra_extra`. What do you think?

arnaudvergnet replied:
That is a good idea as it would remove differences in cfg structure between glossary and variables. The drawback is that it will break compatibility with previous version.

arnaudvergnet replied:
But if we are making breaking changes, might as well use json right away instead of converting between json and config files.

coppolaemilio replied:
We can keep it as is, but I was planning on breaking the compatibility as much as needed now before the actual 1.0 release. The time is now! haha
As you prefer, I'll keep testing this branch and see if there are any errors.

arnaudvergnet replied:
Let's convert definitions to json then, it will be more in line with the rest of the plugin. Only settings and themes will be using cfg.

coppolaemilio replied:
Sounds good to me 👍🏻

arnaudvergnet commented:
@coppolaemilio definitions are now saved in json. Hopefully I did not break anything. I didn't notice any bug but it needs more testing as this change is quite important

arnaudvergnet commented:
While I was at it, I also saved the state (current timeline) in json as well.

coppolaemilio commented:
Sorry I wasn't able to test until now. I'll do the usually test and QA stuff and try to merge soon 👍🏻

coppolaemilio commented:
Starting from an empty project:
![image](https://user-images.githubusercontent.com/2206700/112337755-7cc6a480-8cbe-11eb-819a-70e93b6b3326.png)

arnaudvergnet commented:
As usual I do not test on empty project :) It is fixed

coppolaemilio commented:
Testing an empty project with no variables and just a timeline that has one text event:
![image](https://user-images.githubusercontent.com/2206700/112342589-a386da00-8cc2-11eb-94e8-88a6fdfaff7f.png)

arnaudvergnet commented:
Fixed both bugs!

The first one was because I initialized saves with an empty dictionary. Instead I initialized it with `{'variables': [], 'glossary': []}`, just like I did for default definitions.

As for the dialog not working with the editor node, this was because it returned the file name without `.json`. I changed the start method to allow for timeline name or filename, so I changed how the timeline variable was set with the start method but forgot about the inspector.

All

coppolaemilio commented:
Yes! they look fixed on my end 🎉 I'll merge now and keep looking for stuff. I'll need to really go back over all the changes because it is starting to feel too complex. I would love if we could simplify more stuff after v1.0 is stable
